src/peft/tuners/reft/layer.py

```diff
@@ -0,0 +1,195 @@
+        LycorisLayer.__init__(self, base_layer)
+
+        # ReFT info
+        self.reft_A = nn.ModuleDict({})
+        self.reft_R = nn.ModuleDict({})
+        self.loc = {}
+        self.first_n = {}
+        self.last_n = {}
+        self.dropout = {}
+        base_layer = self.get_base_layer()
+        if isinstance(base_layer, nn.Linear):
+            in_features, out_features = base_layer.in_features, base_layer.out_features
+        elif isinstance(base_layer, nn.Conv2d):
+            in_features, out_features = base_layer.in_channels, base_layer.out_channels
+
+        self.in_features = in_features
+        self.out_features = out_features
+
+
+    @property
+    def _available_adapters(self) -> Set[str]:
+        return {*self.reft_A, *self.reft_R}
+
+    def create_adapter_parameters(self, adapter_name: str, r: int):
+        rotate_layer = torch.nn.Linear(self.out_features, r, bias=False)
+        self.reft_R[adapter_name] = torch.nn.utils.parametrizations.orthogonal(rotate_layer, orthogonal_map='cayley')
+        self.reft_A[adapter_name] =  torch.nn.Linear(self.out_features, r)
+
+    def merge(self, safe_merge: bool = False, adapter_names: Optional[list[str]] = None) -> None:
+        pass
+
+    def unmerge(self) -> None:
+        pass
+
+    def update_layer(
+        self,
+        adapter_name: str,
+        r: int,
+        alpha: float,
+        loc: Optional[str],
+        dropout: float,
+        **kwargs,
+    ) -> None:
+        """Internal function to create loha adapter
+
+        Args:
+            adapter_name (`str`): Name for the adapter to add.
+            r (`int`): Rank for the added adapter.
+            alpha (`float`): Alpha for the added adapter.
+            dropout (`float`): The dropout probability for disabling adapter during training.
+        """
+        if r <= 0:
+            raise ValueError(f"`r` should be a positive integer value but the value passed is {r}")
+
+        self.r[adapter_name] = r
+        first_n, last_n = parse_positions(loc) if loc else (0, 0)
+        self.first_n[adapter_name] = first_n
+        self.last_n[adapter_name] = last_n
+        self.alpha[adapter_name] = alpha
+        self.scaling[adapter_name] = alpha / r
+        self.dropout[adapter_name] = torch.nn.Dropout(dropout)
+
+        # Create weights with provided shape
+        self.create_adapter_parameters(adapter_name, r)
+
+        # Initialize weights
+        if adapter_name in self.reft_A.keys():
+            nn.init.kaiming_uniform_(self.reft_A[adapter_name].weight, a=math.sqrt(5))
+
+        # Move new weights to device
+        weight = getattr(self.get_base_layer(), "weight", None)
+        if weight is not None:
+            # the layer is already completely initialized, this is an update
+            if weight.dtype.is_floating_point or weight.dtype.is_complex:
+                self.to(weight.device, dtype=weight.dtype)
+            else:
+                self.to(weight.device)
```

Comment:
We have a new convenience method for this: `self._move_adapter_to_device_of_base_layer(adapter_name)`.